java/src/org/openqa/selenium/grid/router/HandleSession.java

```diff
@@ -60,92 +60,150 @@
 import org.openqa.selenium.remote.tracing.Status;
 import org.openqa.selenium.remote.tracing.Tracer;
 
-class HandleSession implements HttpHandler, Closeable {
+public class HandleSession implements HttpHandler, Closeable {
 
   private static final Logger LOG = Logger.getLogger(HandleSession.class.getName());
 
-  private static class CacheEntry {
-    private final HttpClient httpClient;
-    private final AtomicLong inUse;
-    // volatile as the ConcurrentMap will not take care of synchronization
-    private volatile Instant lastUse;
-
-    public CacheEntry(HttpClient httpClient, long initialUsage) {
-      this.httpClient = httpClient;
-      this.inUse = new AtomicLong(initialUsage);
-      this.lastUse = Instant.now();
-    }
-  }
-
-  private static class UsageCountingReverseProxyHandler extends ReverseProxyHandler
+  private static class ReverseProxyHandlerCloseable extends ReverseProxyHandler
       implements Closeable {
-    private final CacheEntry entry;
 
-    public UsageCountingReverseProxyHandler(
-        Tracer tracer, HttpClient httpClient, CacheEntry entry) {
+    public ReverseProxyHandlerCloseable(Tracer tracer, HttpClient httpClient) {
       super(tracer, httpClient);
-
-      this.entry = entry;
     }
 
     @Override
     public void close() {
-      // set the last use here, to ensure we have to calculate the real inactivity of the client
-      entry.lastUse = Instant.now();
-      entry.inUse.decrementAndGet();
+      // No operation needed - cache management is handled by Cache builder
     }
   }
 
   private final Tracer tracer;
   private final HttpClient.Factory httpClientFactory;
   private final SessionMap sessions;
-  private final ConcurrentMap<URI, CacheEntry> httpClients;
-  private final ScheduledExecutorService cleanUpHttpClientsCacheService;
+  private final Cache<URI, CacheEntry> httpClientCache;
+
+  /**
+   * Wrapper class to store HttpClient along with its configuration for dynamic cache expiration
+   * based on HttpClient's read timeout.
+   */
+  private static class CacheEntry {
+    private final HttpClient httpClient;
+    private final ClientConfig config;
+    private final long creationTime;
+
+    CacheEntry(HttpClient httpClient, ClientConfig config) {
+      this.httpClient = Require.nonNull("HttpClient", httpClient);
+      this.config = Require.nonNull("ClientConfig", config);
+      this.creationTime = System.currentTimeMillis();
+    }
+
+    HttpClient getHttpClient() {
+      return httpClient;
+    }
+
+    ClientConfig getConfig() {
+      return config;
+    }
+
+    /**
+     * Check if this cache entry has expired based on the HttpClient's read timeout. Method is used
+     * by Cache builder to determine if an entry should be evicted.
+     *
+     * @param lastAccessTime the last time this entry was accessed
+     * @return true if the entry should be evicted
+     */
+    boolean isExpired(long lastAccessTime) {
```

Comment:
I could not find any call to this method, is guess this can be removed or called 😀